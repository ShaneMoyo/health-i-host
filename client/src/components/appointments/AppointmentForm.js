import React, { useState } from 'react';
import { useDispatch } from "react-redux";
import moment from 'moment'
import Select from 'react-select'
import DayPicker from 'react-day-picker';
import TimePicker from 'rc-time-picker';
import { bookAppointment } from './actions';
import { Redirect } from 'react-router-dom';
import 'react-day-picker/lib/style.css';
import 'rc-time-picker/assets/index.css';


export default function AppointmentForm() {

  const [date, setDate] = useState(moment());
  const [type, setType] = useState('massage');
  const [duration, setDuration] = useState(0.5);
  const [status, setStatus] = useState('pending');
  const [appoitmentBooked, setAppoitmentBooked] = useState(false);
  const dispatch = useDispatch();


  const onDateChange = newDate => {
    if (!newDate) return;
    newDate = moment(newDate);
    const day = newDate.date();
    const month = newDate.month();
    setDate(date.date(day).month(month));
  }

  const onTimeChange = time => {
    if (!time) return;
    time = moment(time);
    const day = time.hour();
    const month = time.minute();
    setDate(date.hour(day).minute(month));

  }

  const handleSubmit = () => {
    return dispatch(bookAppointment({ date, type, status, duration })).then(() => setAppoitmentBooked(!appoitmentBooked))
  }

  const disabledHours = () => [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 20, 21, 22, 23];


  const typeOptions = [
    { value: 'massage', label: 'Massage Therapy' },
    { value: 'mineral', label: 'Mineral Consultation' },
    { value: 'movement', label: 'Movement' }
  ];
  const durationOptions = [
    { value: 0.5, label: '30 Minutes' },
    { value: 1, label: '1 Hour' },
    { value: 1.5, label: '1 Hour 30 Minutes' },
    { value: 2, label: '2 Hours' }
  ];

  if (appoitmentBooked) { return <Redirect to='/appointment/booked'/> }

  return (
        <div class="container has-text-centered">
          <br/><br/>
          <section class="hero is-warning is-fullheight">
            <div class="column is-waring is-one-third is-offset-one-third">
              <div class="box animated fadeIn is-warning" >
                <div>
                  <div class="field">
                    <Select defaultValue={typeOptions[0]} options={typeOptions} onChange={({ value }) => setType(value)}/>
                  </div>

                  <div class="field">
                    <Select defaultValue={durationOptions[1]} options={durationOptions} onChange={({ value }) => setDuration(value)}/>
                  </div>

                  <div class="field">
                    <div calss="container">
                      <DayPicker onDayClick={onDateChange}/>
                    </div>
                  </div>
                  <div class="field">
                    <TimePicker
                    showSecond={false}
                    hideDisabledOptions={true}
                    defaultValue={date}
                    className="xxx"
                    onChange={onTimeChange}
                    disabledHours={disabledHours}
                    format='h:mm a'
                    minuteStep={30}
                    use12Hours
                    inputReadOnly
                    />
                  </div>
                  <hr/>
                  <div class="field">
                    <label class="label"></label>
                    <button class="button is-medium is-info" onClick={handleSubmit}>Book Appointment</button>
                  </div>
                </div>
              </div>
            </div>
          </section>
        </div>
  );
}
