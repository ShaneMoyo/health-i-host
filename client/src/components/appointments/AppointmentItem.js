import React, { useState } from 'react';
import { useDispatch, useSelector } from "react-redux";
import { updateAppointment } from './actions';
import moment from 'moment'
import DateTimePicker from 'react-datetime-picker';
import Fade from 'react-reveal/Fade';
import Icon from '@mdi/react'
import {  mdiCloseCircleOutline, mdiDotsHorizontal, mdiCommentProcessingOutline, mdiPencilOutline, mdiWindowClose, mdiWrenchOutline, mdiPlusBoxOutline, mdiPlus, mdiCheck } from '@mdi/js'
import classNames from 'classnames';
import PropTypes from 'prop-types';

export default function AppointmentItem (props){

  const { appointment, toggleModal } = props;
  const { note: initialNote, date: initialDate, status, type, _id } = appointment;
  const [note, setNote] = useState(initialNote);
  const [addNote, setaAddNote] = useState(false);
  const [edit, setEdit] = useState(false);
  const showNotes = initialNote && !addNote;
  const loading = useSelector(state => state.loading);
  const dispatch = useDispatch();

  const handleClientUpdate = () => {
    return dispatch(updateAppointment({ ...appointment, note }))
      .then(()=> { setaAddNote(false) })
  }

  const capatilizedType = appointment.type.charAt(0).toUpperCase() + appointment.type.slice(1);

  const colorMap = {
    massage: 'success',
    movement: 'primary',
    mineral: 'warning'
  }
  const durationOptions = {
    0.5: '30 Minutes',
    1: '1 Hour',
    1.5: '1 Hour 30 Minutes',
    2: '2 Hours'
  };
  const statusMap = {
    'pending': 'Pending confirmation'
  }
  const color = colorMap[appointment.type];
  const className = `tile is-child notification has-text-white is-${color}`

  return <li class="appointments">
          <Fade>
            <div>

              <div class={classNames('tile is-parent is-vertical grow', { more: edit })}>
                <article class={className}>

                  <span class="icon has-text-info is-pulled-right " onClick={() => setEdit(!edit)}>
                    <i class="fas fa-info-circle">
                      <Icon
                        className="animated fadeIn"
                        path={ edit ? mdiWindowClose : mdiDotsHorizontal}
                        title="User Profile"
                        size={1}
                        horizontal
                        vertical
                        color="white"
                        />
                    </i>
                  </span>

                  <br/>
                  <p class="title">{capatilizedType}</p>
                  <p class="subtitle">
                    {moment(appointment.date).format('M/DD h:mm a')}
                  </p>
                  { edit &&
                    <div>
                    <hr class="hr-white animated fadeIn"/>
                    <p class="subtitle animated fadeInDown">
                      {durationOptions[appointment.duration]}
                    </p>
                    <p class="subtitle animated fadeInDown">{statusMap[appointment.status]}</p>
                    <div class="is-divider" data-content="OR"></div>

                    { addNote &&
                      <div class="control animated fadeIn">
                        <span class="icon is-small is-pulled-right" onClick={() => setaAddNote(!addNote)}>
                          <Icon path={mdiWindowClose} size={0.7} color="white"/>
                        </span>
                        <textarea class="textarea is-outlined" placeholder={appointment.note} onChange={({ target }) => setNote(target.value)}></textarea>
                        <br/>
                      </div>
                    }

                    { showNotes && <p class="subtitle animated fadeInDown">Note - {appointment.note}</p> }

                    <p class="buttons animated fadeInDown">
                      { !addNote ?
                        <a class="button is-primary is-inverted is-outlined" onClick={() => setaAddNote(!addNote)}>
                          <span>{ initialNote ? 'Edit Note' : 'Add Note' }</span>
                          <span class="icon is-small">
                            <Icon path={mdiPlus} size={0.7} color="white"/>
                          </span>
                        </a> :
                        <a class="button is-primary is-inverted is-outlined" onClick={() => handleClientUpdate()}>
                          <span>Save</span>
                          <span class="icon is-small">
                            <Icon path={mdiCheck} size={0.7} color="white"/>
                          </span>
                        </a>

                      }

                      <a class="button is-primary is-inverted is-outlined" onClick={() => toggleModal(_id)}>
                        <span>Cancel</span>
                        <span class="icon is-small">
                          <Icon path={mdiWindowClose} size={0.7} color="white"/>
                        </span>
                      </a>

                    </p>
                    </div>
                  }


                </article>
              </div>
            </div>
          </Fade>
        </li>
}

AppointmentItem.propTypes = {
  appointment: PropTypes.object,
  toggleModal: PropTypes.func
}
