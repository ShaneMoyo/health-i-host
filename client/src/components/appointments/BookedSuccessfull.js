import React from 'react';
import { NavLink } from 'react-router-dom';

const NavBarLink = props => <NavLink {...props} className="nav-link" activeClassName="active"/>;
function AppointmentForm() {
    return (
          <div class="container has-text-centered">
            <br/><br/>
            <section class="hero is-warning is-fullheight">
              <div class="column is-waring is-6 is-offset-3">
                <div class="box animated fadeIn is-warning" >
                  <div>
                    <div class="animated fadeIn title is-6">Appointment Booked Succesfully</div>
                    <br/>
                    <div class="animated fadeIn button is is-info"><NavBarLink exact to="/appointment/me">Proceed to My Appointments</NavBarLink></div>
                  </div>
                </div>
              </div>
            </section>
          </div>
    );
}

export default AppointmentForm;
