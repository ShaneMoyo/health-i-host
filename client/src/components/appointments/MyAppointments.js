import React, { useState, useEffect } from 'react';
import { useDispatch, useSelector } from "react-redux";
import { loadMyAppointments, deleteAppointment } from './actions';
import AppointmentItem from './AppointmentItem';
import CancelModal from '../utils/CancelModal.js';
import moment from 'moment'
import Fade from 'react-reveal/Fade';
import Icon from '@mdi/react'
import {  mdiDotsHorizontal, mdiPlus } from '@mdi/js'
import { NavLink } from 'react-router-dom';
import "react-loader-spinner/dist/loader/css/react-spinner-loader.css"
import Loader from 'react-loader-spinner'

const NavBarLink = props => <NavLink {...props} className="nav-link" activeClassName="active"/>;

export default function MyAppointments () {

  async function loadAppointments() { await dispatch(loadMyAppointments()) }

  useEffect(() => { loadAppointments() }, []);

  const loading = useSelector(state => state.loading);
  const appointments = useSelector(state => state.appointments);
  const dispatch = useDispatch();

  const [showModal, setShowModal] = useState(false);
  const [deleteId, setDeleteId] = useState(null);
  const handelToggleModdle = (deleteId) => {
    setShowModal(!showModal);
    setDeleteId(deleteId)
  }
  const handleDeleteAppointment = () => dispatch(deleteAppointment(deleteId)).then(() => setShowModal(false))
  const sortedAppointments = appointments.sort((a, b) => a.date - b.date);
  const myAppointments = sortedAppointments.map(appointment => <AppointmentItem key={appointment._id} appointment={ appointment } toggleModal={() => handelToggleModdle(appointment._id) }/>)

  return(
      <div class="hero-body">
        <div class="container">
          <div class="column is-6 is-offset-3">
            <div class="box animated fadeIn" >
              <CancelModal toggleModal={setShowModal}
                showCancelModal={showModal}
                handleDeleteAppointment={handleDeleteAppointment}
              />
              { loading ?
                <div class="has-text-centered">
                <Loader
                   type="Grid"
                   color="#8ab6c6"
                   height={500}
                   width={100}
                /></div>:
                <ul class="appointment-list">
                  {myAppointments}
                  <li class="appointments">
                  <NavBarLink exact to="/appointment">
                    <div class="tile is-parent is-vertical grow more">
                      <article class="tile is-child notification has-text-white is-info">
                      <p class="title">
                        Book
                        <span class="icon has-text-info is-pulled-right ">
                        <i class="fas fa-info-circle">
                        <Icon
                        className="animated fadeIn"
                        path={mdiPlus}
                        title="User Profile"
                        size={1}
                        horizontal
                        vertical
                        color="white"
                        />
                        </i>
                        </span>
                      </p>
                      </article>
                    </div>
                    </NavBarLink>
                  </li>
                </ul>
              }
            </div>
          </div>
        </div>
      </div>
  );
}
