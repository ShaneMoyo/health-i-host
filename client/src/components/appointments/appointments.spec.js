import React from 'react';
import { shallow } from 'enzyme';
import AppointmentForm from './AppointmentForm';
import AppointmentItem from './AppointmentItem';
import BookedSuccessfull from './BookedSuccessfull';
import MyAppointments from './MyAppointments';

describe('AppointmentForm', () => {
  it('should render correctly AppointmentForm component', () => {
    const component = shallow(<AppointmentForm />);

    expect(component).toMatchSnapshot();
  });

  it('should render correctly AppointmentItem component', () => {
    const component = shallow(<AppointmentItem />);

    expect(component).toMatchSnapshot();
  });

  it('should render correctly MyAppointments component', () => {
    const component = shallow(<MyAppointments />);

    expect(component).toMatchSnapshot();
  });

  it('should render correctly BookedSuccessfull component', () => {
    const component = shallow(<BookedSuccessfull />);

    expect(component).toMatchSnapshot();
  });
});
