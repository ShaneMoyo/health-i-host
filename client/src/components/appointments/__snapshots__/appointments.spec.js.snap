// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentForm should render correctly BookedSuccessfull component 1`] = `
<div
  class="container has-text-centered"
>
  <br />
  <br />
  <section
    class="hero is-warning is-fullheight"
  >
    <div
      class="column is-waring is-6 is-offset-3"
    >
      <div
        class="box animated fadeIn is-warning"
      >
        <div>
          <div
            class="animated fadeIn title is-6"
          >
            Appointment Booked Succesfully
          </div>
          <br />
          <div
            class="animated fadeIn button is is-info"
          >
            <NavBarLink
              exact={true}
              to="/appointment/me"
            >
              Proceed to My Appointments
            </NavBarLink>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
